### PR TITLE
Basic Testing

### PR DESCRIPTION
**Pull request by surface74**
1. [Basic Testing](https://github.com/AlreadyBored/basic-testing)
2. [Repository](https://github.com/surface74/basic-testing)
3. Done 17.10.2024 / deadline 22.10.2024
4. Score: 258/258

### CODE REVIEW

File: src/04-test-class/index.test.ts [hunk 1]
```diff
@@ -1,44 +1,106 @@
...
+    expect(account2.getBalance()).toBe(money2 + amount);
   });
 
   test('fetchBalance should return number in case if request did not failed', async () => {
-    // Write your tests here
+    const money = 100;
+    const account = getBankAccount(money);
+
+    const balance = await account.fetchBalance();
+    expect(typeof balance === 'number' || balance === null).toBeTruthy();
```
Comment (Nurgissa) [inline src/04-test-class/index.test.ts:79]: In this test you supposed to mock `lodash/random` function and make tests determenistic.

File: src/04-test-class/index.test.ts [hunk 2]
```diff
@@ -1,44 +1,106 @@
...
   });
 
   test('should set new balance if fetchBalance returned number', async () => {
-    // Write your tests here
+    const money = 100;
+    const account = getBankAccount(0);
+
+    const mockFetchBalance = jest.fn();
+    mockFetchBalance.mockImplementation(async () => money);
+    account.fetchBalance = mockFetchBalance;
```
Comment (Nurgissa) [inline src/04-test-class/index.test.ts:88]: You should not mock `account.fetchBalance` in this test. Instead, you should test implementation of `account.fetchBalance` which relies on `lodash/random`.

In this test you supposed to mock `lodash/random` function and make tests determenistic.

File: src/04-test-class/index.test.ts [hunk 3]
```diff
@@ -1,44 +1,106 @@
...
+    expect(account.getBalance()).toBe(money);
   });
 
   test('should throw SynchronizationFailedError if fetchBalance returned null', async () => {
-    // Write your tests here
+    const account = getBankAccount(0);
+
+    const mockFetchBalance = jest.fn();
+    mockFetchBalance.mockImplementation(async () => null);
+    account.fetchBalance = mockFetchBalance;
```
Comment (Nurgissa) [inline src/04-test-class/index.test.ts:100]: You should not mock `account.fetchBalance` in this test. Instead, you should test implementation of `account.fetchBalance` which relies on `lodash/random`.

In this test you supposed to mock `lodash/random` function and make tests determenistic.

File: src/06-mocking-node-api/index.test.ts [hunk 4]
```diff
@@ -11,11 +11,29 @@ describe('doStuffByTimeout', () => {
...
+
+    expect(setTimeout).toHaveBeenCalledTimes(1);
+    expect(setTimeout).toHaveBeenCalledWith(callback, timeout);
   });
 
   test('should call callback only after timeout', () => {
-    // Write your test here
+    const timeout = 1000;
+    const callback = jest.fn();
+    jest.spyOn(global, 'setTimeout');
```
Comment (Nurgissa) [inline src/06-mocking-node-api/index.test.ts:27]: Do not spy but mock the implementation.

File: src/06-mocking-node-api/index.test.ts [hunk 5]
```diff
@@ -29,24 +47,56 @@ describe('doStuffByInterval', () => {
   });
 
   test('should set interval with provided callback and timeout', () => {
-    // Write your test here
+    const timeout = 1000;
+    const callback = jest.fn();
+    jest.spyOn(global, 'setInterval');
```
Comment (Nurgissa) [inline src/06-mocking-node-api/index.test.ts:52]: Do not spy but mock implementation.

File: src/06-mocking-node-api/index.test.ts [hunk 6]
```diff
@@ -29,24 +47,56 @@ describe('doStuffByInterval', () => {
...
   test('should call join with pathToFile', async () => {
-    // Write your test here
+    const join = jest.spyOn(path, 'join');
+    const existFile = './index.ts';
+    await readFileAsynchronously(existFile);
+
+    expect(join).toHaveBeenCalledWith(expect.any(String), existFile);
   });
 
   test('should return null if file does not exist', async () => {
```
Comment (Nurgissa) [inline src/06-mocking-node-api/index.test.ts:86]: This test is not correct.

File: src/06-mocking-node-api/index.test.ts [hunk 7]
```diff
@@ -11,11 +11,29 @@ describe('doStuffByTimeout', () => {
   });
 
   test('should set timeout with provided callback and timeout', () => {
-    // Write your test here
+    const timeout = 1000;
+    const callback = jest.fn();
+    jest.spyOn(global, 'setTimeout');
```
Comment (Nurgissa) [inline src/06-mocking-node-api/index.test.ts:16]: Do not spy, but mock the implementation.

File: src/06-mocking-node-api/index.test.ts [hunk 8]
```diff
@@ -29,24 +47,56 @@ describe('doStuffByInterval', () => {
...
+    const join = jest.spyOn(path, 'join');
+    const existFile = './index.ts';
+    await readFileAsynchronously(existFile);
+
+    expect(join).toHaveBeenCalledWith(expect.any(String), existFile);
   });
 
   test('should return null if file does not exist', async () => {
-    // Write your test here
+    const absentFile = './index';
```
Comment (Nurgissa) [inline src/06-mocking-node-api/index.test.ts:87]: deducting 3 pts.

File: src/06-mocking-node-api/index.test.ts [hunk 9]
```diff
@@ -1,5 +1,5 @@
-// Uncomment the code below and write your tests
-// import { readFileAsynchronously, doStuffByTimeout, doStuffByInterval } from '.';
+import path from 'path';
+import { doStuffByTimeout, doStuffByInterval, readFileAsynchronously } from '.';
```
Comment (Nurgissa) [inline src/06-mocking-node-api/index.test.ts:3]: In this one should mock nodejs build-in libraries. I do not see any mocking except jest spying

File: src/04-test-class/index.test.ts (hunk 3 above)
Comment (Nurgissa) [reply src/04-test-class/index.test.ts:100]: Deducting 3 points

File: src/04-test-class/index.test.ts (hunk 2 above)
Comment (Nurgissa) [reply src/04-test-class/index.test.ts:88]: Deducting 3 points

File: src/04-test-class/index.test.ts (hunk 1 above)
Comment (Nurgissa) [reply src/04-test-class/index.test.ts:79]: Deducting 3 points

File: src/06-mocking-node-api/index.test.ts (hunk 6 above)
Comment (Nurgissa) [reply src/06-mocking-node-api/index.test.ts:86]: For example,
image
Deducting 3 points.

File: src/06-mocking-node-api/index.test.ts [hunk 10]
```diff
@@ -29,24 +47,56 @@ describe('doStuffByInterval', () => {
...
-    // Write your test here
+    const existFile = './index.ts';
+    const absentFile = './index';
+
+    const existContent = await readFileAsynchronously(existFile);
+    const nullContent = await readFileAsynchronously(absentFile);
+
+    expect(
+      typeof existContent === 'string' && nullContent === null,
+    ).toBeTruthy();
```
Comment (Nurgissa) [inline src/06-mocking-node-api/index.test.ts:100]: This test is partially correct. You should mock nodejs library instead providing non-exisiting or existing file. If you change `absentFile` to `./index.ts`, your test will fail.

image

Comment (Nurgissa) [reply src/06-mocking-node-api/index.test.ts:100]: Deducting 3 points.

File: src/04-test-class/index.test.ts (hunk 1 above)
Comment (surface74) [reply src/04-test-class/index.test.ts:79]: I`ve got it

File: src/04-test-class/index.test.ts (hunk 2 above)
Comment (surface74) [reply src/04-test-class/index.test.ts:88]: OK

File: src/04-test-class/index.test.ts (hunk 3 above)
Comment (surface74) [reply src/04-test-class/index.test.ts:100]: In this case I'm disagree. The task is `should throw SynchronizationFailedError if fetchBalance returned null`. So I can mock fetchBalance to return null.

File: src/06-mocking-node-api/index.test.ts (hunk 6 above)
Comment (surface74) [reply src/06-mocking-node-api/index.test.ts:86]: I got it.

File: src/06-mocking-node-api/index.test.ts (hunk 10 above)
Comment (surface74) [reply src/06-mocking-node-api/index.test.ts:100]: I got it

File: src/07-mocking-lib-api/index.test.ts [hunk 11]
```diff
@@ -1,17 +1,40 @@
...
-    // Write your test here
+    const requestedUrl = 'todos';
+    const data = await throttledGetDataFromApi(requestedUrl);
+    expect(data).toBe(requestedUrl);
   });
 
   test('should return response data', async () => {
-    // Write your test here
+    const data = await throttledGetDataFromApi('todos');
+    expect(data).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (surface74) [reply src/07-mocking-lib-api/index.test.ts:38]: OK

**Pull request merged**